pkg/storage/azure/azure.go

```diff
@@ -0,0 +1,506 @@
+	}
+}
+
+func (d *driver) getAzureConfig() (*clusterconfig.Azure, error) {
+	cfg, err := clusterconfig.GetAzureConfig(d.Listers)
+	if err != nil {
+		return nil, err
+	}
+	return cfg, nil
+}
+
+func (d *driver) storageAccountsClient(cfg *clusterconfig.Azure) (storage.AccountsClient, error) {
+	auth, err := auth.NewClientCredentialsConfig(cfg.ClientID, cfg.ClientSecret, cfg.TenantID).Authorizer()
+	if err != nil {
+		return storage.AccountsClient{}, err
+	}
+
+	storageAccountsClient := storage.NewAccountsClient(cfg.SubscriptionID)
+	storageAccountsClient.Authorizer = auth
+	storageAccountsClient.AddToUserAgent(parameters.UserAgent)
+
+	return storageAccountsClient, nil
+}
+
+// ConfigEnv configures the environment variables that will be used in the
+// image registry deployment.
+func (d *driver) ConfigEnv() (envs []corev1.EnvVar, err error) {
+	envs = append(envs,
+		corev1.EnvVar{Name: "REGISTRY_STORAGE", Value: "azure"},
+		corev1.EnvVar{Name: "REGISTRY_STORAGE_AZURE_CONTAINER", Value: d.Config.Container},
+		corev1.EnvVar{Name: "REGISTRY_STORAGE_AZURE_ACCOUNTNAME", Value: d.Config.AccountName},
+		corev1.EnvVar{
+			Name: "REGISTRY_STORAGE_AZURE_ACCOUNTKEY",
+			ValueFrom: &corev1.EnvVarSource{
+				SecretKeyRef: &corev1.SecretKeySelector{
+					LocalObjectReference: corev1.LocalObjectReference{
+						Name: imageregistryv1.ImageRegistryPrivateConfiguration,
+					},
+					Key: "REGISTRY_STORAGE_AZURE_ACCOUNTKEY",
+				},
+			},
+		},
+	)
+	return
+}
+
+func (d *driver) Volumes() ([]corev1.Volume, []corev1.VolumeMount, error) {
+	return nil, nil, nil
+}
+
+// Secrets returns a map of the storage access secrets.
+func (d *driver) Secrets() (map[string]string, error) {
+	cfg, err := d.getAzureConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	key := cfg.AccountKey
+	if !cfg.UPI {
+		storageAccountsClient, err := d.storageAccountsClient(cfg)
+		if err != nil {
+			return nil, err
+		}
+
+		key, err = getAccountPrimaryKey(storageAccountsClient, cfg.ResourceGroup, d.Config.AccountName)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return map[string]string{
+		"REGISTRY_STORAGE_AZURE_ACCOUNTKEY": key,
+	}, nil
+}
+
+// StorageExists checks if the storage container exists and accessible.
+func (d *driver) StorageExists(cr *imageregistryv1.Config) (bool, error) {
+	if d.Config.AccountName == "" || d.Config.Container == "" {
+		util.UpdateCondition(cr, imageregistryv1.StorageExists, operatorapiv1.ConditionFalse, storageExistsReasonNotConfigured, "Storage is not configured")
+		return false, nil
+	}
+
+	cfg, err := d.getAzureConfig()
+	if err != nil {
+		util.UpdateCondition(cr, imageregistryv1.StorageExists, operatorapiv1.ConditionUnknown, storageExistsReasonConfigError, fmt.Sprintf("Unable to get configuration: %s", err))
+		return false, err
+	}
+
+	if cfg.UPI {
+		util.UpdateCondition(cr, imageregistryv1.StorageExists, operatorapiv1.ConditionTrue, storageExistsReasonUserManaged, "Storage is managed by the user")
+		return true, nil
+	}
```

Comment:
Even for UPI, shouldn't this method verify that the named storage container exists? Or is that not possible with the way Azure's account/permission model works?